flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java

```diff
@@ -241,6 +243,22 @@ public void putJobGraph(JobGraph jobGraph) throws Exception {
         LOG.info("Added {} to {}.", jobGraph, jobGraphStateHandleStore);
     }
 
+    @Override
+    public void putJobResourceRequirements(
+            JobID jobId, JobResourceRequirements jobResourceRequirements) throws Exception {
+        synchronized (lock) {
+            @Nullable final JobGraph jobGraph = recoverJobGraph(jobId);
+            if (jobGraph == null) {
+                throw new FileNotFoundException(
+                        String.format(
+                                "JobGraph for job [%s] was not found in JobGraphStore and is needed for attaching JobResourceRequirements.",
+                                jobId));
+            }
+            JobResourceRequirements.writeToJobGraph(jobGraph, jobResourceRequirements);
+            putJobGraph(jobGraph);
```

Comment:
Could you explain how do we prevent concurrent writes?
I'm concerned about a case when a stand-by JM runs through the same code path.

IIUC, that was not possible before, because it was called only during job submission. But now it will also be called for existing jobs.

Reply:
A contender JM shouldn't do any writes.

Reply:
Per my understanding, in the worst case, a contender JM **_might_** discover that it's not the leader anymore while adding checkpoint to ZK/etcd.

CheckpointCounters use `getAndIncrement` for new checkpoints.

Reply:
That can indeed happen. It's unlikely to be a _problem_ though, because even in the worst case (i.e., a JM is writing the requirements while loosing leadership, and another one gains leadership and starts reading job graphs before that write is complete, which is already unlikely) you we might "forget" the last set requirements after a failover.

Reply:
I think you're right, the only impact will be overwriting resource requirements, which is acceptable.